velox/substrait/SubstraitUtils.cpp

```diff
@@ -0,0 +1,160 @@
+      for (const auto& type : sTypes) {
+        parseType(type);
+      }
+      break;
+    }
+    case ::substrait::Type::KindCase::kString: {
+      typeName = "STRING";
+      nullability = sType.string().nullability();
+      break;
+    }
+    default:
+      VELOX_NYI("Substrait parsing for type {} not supported.", typeName);
+  }
+  bool nullable;
+  switch (nullability) {
+    case ::substrait::Type_Nullability::
+        Type_Nullability_NULLABILITY_UNSPECIFIED:
+      nullable = true;
+      break;
+    case ::substrait::Type_Nullability::Type_Nullability_NULLABILITY_NULLABLE:
+      nullable = true;
+      break;
+    case ::substrait::Type_Nullability::Type_Nullability_NULLABILITY_REQUIRED:
+      nullable = false;
+      break;
+    default:
+      VELOX_NYI(
+          "Substrait parsing for nullability {} not supported.", nullability);
+  }
+  std::shared_ptr<SubstraitType> substraitType =
+      std::make_shared<SubstraitType>(typeName, nullable);
+  return substraitType;
+}
+
+std::vector<std::shared_ptr<SubstraitParser::SubstraitType>>
+SubstraitParser::parseNamedStruct(const ::substrait::NamedStruct& namedStruct) {
+  const auto& sNames = namedStruct.names();
+  std::vector<std::string> nameList;
+  nameList.reserve(sNames.size());
+  for (const auto& sName : sNames) {
+    nameList.emplace_back(sName);
+  }
+  // Parse Struct.
+  const auto& sStruct = namedStruct.struct_();
+  const auto& sTypes = sStruct.types();
+  std::vector<std::shared_ptr<SubstraitParser::SubstraitType>>
+      substraitTypeList;
+  substraitTypeList.reserve(sTypes.size());
+  for (const auto& type : sTypes) {
+    substraitTypeList.emplace_back(parseType(type));
+  }
+  return substraitTypeList;
+}
+
+int32_t SubstraitParser::parseReferenceSegment(
+    const ::substrait::Expression::ReferenceSegment& sRef) {
+  auto typeCase = sRef.reference_type_case();
+  switch (typeCase) {
+    case ::substrait::Expression::ReferenceSegment::ReferenceTypeCase::
+        kStructField: {
+      auto sField = sRef.struct_field();
+      auto fieldId = sField.field();
+      return fieldId;
+    }
+    default:
+      VELOX_NYI(
+          "Substrait conversion not supported for ReferenceSegment '{}'",
+          typeCase);
+  }
+}
+
+std::vector<std::string> SubstraitParser::makeNames(
+    const std::string& prefix,
+    int size) {
+  std::vector<std::string> names;
+  names.reserve(size);
+  for (int i = 0; i < size; i++) {
+    names.emplace_back(fmt::format("{}_{}", prefix, i));
+  }
+  return names;
+}
+
+std::string SubstraitParser::makeNodeName(int node_id, int col_idx) {
+  return fmt::format("n{}_{}", node_id, col_idx);
+}
+
+std::string SubstraitParser::findSubstraitFunction(
+    const std::unordered_map<uint64_t, std::string>& functionMap,
+    const uint64_t& id) const {
+  if (functionMap.find(id) == functionMap.end()) {
+    VELOX_FAIL("Could not find function id {} in function map.", id);
+  }
+  std::unordered_map<uint64_t, std::string>& map =
```

Comment:
do you actually want to add the key if it doesn't exist in here? Look like you should call .find() on the map, then return a default value (or throw) if it doesn't exist?

Reply:
Thanks for your comment.
The function map is pre-constructed in the Substrait plan. If the key doesn't exist, then the function map should be wrongly constructed or the key is invalid. Therefore, an exception is thrown in this case. The function map is decided on the Substrait plan producer side, so the unfounded keys should not be added here.